grails-app/services/au/org/ala/collectory/IptService.groovy

```diff
@@ -119,11 +119,14 @@ class IptService {
                                 }
                             }
 
-                            def emails = old.getContacts().collect { it.contact.email }
-
                             //sync contacts
                             update.contacts.each { contact ->
-                                if (!emails.contains(contact.email)) {
+                                def existingContact = old.getContacts().find {
+                                    (it.contact.email && !it.contact.email.isEmpty() && it.contact.email == contact.email) ||
+                                            (it.contact.firstName == contact.firstName && it.contact.lastName == contact.lastName)
```

Comment:
I am not familiar with the context so this might be misinformed.

The firstName/lastName comparison may be a problem. For example:
1. A contact has a new email (same firstName and lastName). This no longer adds a new contact with the new email. I would always want to add the new email.
2. A contact has same firstName and lastName as an old contact. The new contact is not added. While unlikely, I do not want prevent two contacts with the same name from existing.

Reply:
Yes, it's problematic, but as email is not mandatory and we don't have a primary key, we should use the names.

For instance, this is the IPT:
![image](https://github.com/AtlasOfLivingAustralia/collectory/assets/180085/5c5dc762-6ec3-4bb4-b328-e94049d923bc)
Duplicated contacts but with different names or emails results in duplicate contacts. The IPT only fix these duplicates if you use exactly the same names and email.

![image](https://github.com/AtlasOfLivingAustralia/collectory/assets/180085/9cf1c9aa-e29b-426b-b645-3ecda93888d2)

I was thinking that probably a better solution for us to maintain contacts in sync, is to always remove all `contact_for` from a data resource and re-add with the new contacts instead of trying to compare and update. In this case we can have "orphans" old contacts without any dr, so maybe we can also check this and remove the orphans contacts (imagine a contact without email or a misspelled name that is corrected in a new version of a dr).

I was thinking of adding in a future PR and doing this PR something similar to what we have but improved and less drastic and easy to review.

What do you think?

Reply:
For this PR, when merging new information, it must not exclude more new contacts that it already does. I now understand it is already excluding those without an email or those that share an email (given that it is not the primary key). 

At the very least, this name comparison line needs a change. I'll leave it to you if you want to address it in this PR or remove this line and address it a new PR.

I agree with your suggestion. It is preferable to:
1. Add all new contacts. There will be a comparison of all contact fields to determine what is new.
2. Remove old contacts (the ContactFor) that are no longer in the IPT source. This makes the assumption no contact information, for an IPT resource, will be manually altered in the collectory.
3. Deleting of contacts that are orphaned (those without a ContactFor)

Having looked at the code a little longer I have another question. Previously the creator contact was added as a `primary=true`. The additional `dataset.contacts` and `dataset.associatedParties` are also added as `primary=true`. Should these be `primary=false`?

Reply:
Currently we have a list of primary contacts and we check if it should be marked as primary or not. See below the part: 
`boolean isPrimaryContact = update.primaryContacts.contains(contact)`